feat: Use itwinui-variables

mayank99 opened the pull request:
Added `@itwin/itwinui-variables` import in html pages. Changed theming logic to use `data-iui-theme` attribute, both in `<theme-button>` component as well as backstop tests.

Major changes made:
- [x] **Spacing**
- [x] **Text and icon colors**
- [x] **Baseline**: `$iui-baseline` replaced with `--iui-size-s`, `$iui-line-height` with `--iui-size-l`.
- [x] **Typography**: _Had to fix `--iui-font-weight-semibold` to be 600 instead of 500._ 
- [x] **Shadows**
- [x] **Border-radii**
- [x] Removed files with old sass variables wherever applicable
- [x] **Speed variables**: Added new `duration.scss` in variables pkg.
- [x] **Component height/offset**: _Replaced with existing variables and also added new component-height vars._
- [x] Transparency (existing usage of `-rgb` + opacity): _This required some non 1:1 changes/rewrites in a few places._
- [x] Icon vars removed
- [x] **Badge**: Previously was using specific colors from other themes. The values are now hardcoded.

---

_See deploy preview at https://itwin.github.io/iTwinUI/718/_

mayank99 commented:
I think this is now ready for review. The only thing pending is to potentially add a few more CSS variables for the colors that we are not happy with.

I have not updated the test images yet because I would like to get the color changes in first.

FlyersPh9 commented on packages/itwinui-css/src/button/split-button.scss:
```diff
@@ -31,7 +30,7 @@
 
   :where(.iui-button:last-child) {
-    --_iui-button-padding-inline: #{$iui-xs};
+    --_iui-button-padding-inline: var(--iui-size-2xs);
```
```suggestion
    --_iui-button-padding-inline: var(--iui-size-2xs);
    
    min-width: unset;
```
We need to remove the min-width on the 2nd button within the split button.

How it looks currently:
![Screen Shot 2022-08-12 at 10 00 34 AM](https://user-images.githubusercontent.com/849817/184369291-8f2a55af-4c30-4e18-ade3-3ec055816410.png)

How it should look:
![Screen Shot 2022-08-12 at 10 01 36 AM](https://user-images.githubusercontent.com/849817/184369479-a0bd5bf8-f94e-4f06-bae7-20aa6e3ca0f8.png)

FlyersPh9 commented on packages/itwinui-css/src/color-picker/color-picker.scss:
```diff
@@ -73,12 +72,12 @@ $iui-hover-box-shadow: 0 0 1px $iui-xxs + 1 rgba(var(--iui-color-foreground-body
-  gap: $iui-s;
+  min-height: var(--iui-size-l);
+  min-width: var(--iui-size-l);
+  gap: var(--iui-size-xs);
```
```suggestion
  gap: var(--iui-size-xs);
  align-items: center;
```
The borderless button & swatches are not aligned in this branch:
![Screen Shot 2022-08-12 at 10 12 00 AM](https://user-images.githubusercontent.com/849817/184371479-655c015e-47e4-4f8c-8339-96e91522acea.png)

mayank99 replied:
Fixed

FlyersPh9 commented:
I'm seeing a visual bug with the dropdown icon in the small select within the information panel demo:

![Screen Shot 2022-08-12 at 10 23 05 AM](https://user-images.githubusercontent.com/849817/184373693-b2e76218-4db1-46d3-860d-98f8e32f5fbc.png)

FlyersPh9 commented:
All instances of `@include iui-icon-style('3xl')` need to have the `'` removed.  Tagging @elephantcatdog just so she is aware because I think these changes came from #701. Example:

```diff
- @include iui-icon-style('3xl');
+ @include iui-icon-style(3xl);
```

FlyersPh9 commented:
I noticed local & live both list **index** as a demo page & it just refreshes the homepage.  Not sure when this got introduced.

![Screen Shot 2022-08-12 at 10 56 43 AM](https://user-images.githubusercontent.com/849817/184382372-4c0fad11-73f6-4ed3-8b6a-5f43961b7bb1.png)

FlyersPh9 commented on packages/itwinui-css/src/table/table.scss:
```diff
@@ -154,11 +154,7 @@
-          rgba(var(--iui-color-foreground-primary-rgb), var(--iui-opacity-6))
-        ),
-          linear-gradient(var(--iui-color-background-1), var(--iui-color-background-1));
+        background-color: var(--iui-color-background-subtle-primary);
```
```suggestion
        background-color: var(--iui-color-background-transparent-hover);
```
To align with menu.

FlyersPh9 commented on packages/itwinui-css/src/progress-indicator/overlay.scss:
```diff
@@ -15,17 +15,10 @@
-    background-color: rgba(var(--iui-color-background-1-rgb), var(--iui-opacity-3));
-    backdrop-filter: $iui-blur-filter;
-  }
+  @include iui-blur($opacity: 3);
```
This needs to be looked into:
![Screen Shot 2022-08-12 at 11 55 45 AM](https://user-images.githubusercontent.com/849817/184395224-2bf7ee53-e5f2-4bc6-9a55-9f502686e344.png)

mayank99 replied:
Fixed by adding a `-background-1-hsl` variable.

FlyersPh9 commented on packages/itwinui-css/src/side-navigation/side-navigation.scss:
```diff
@@ -124,7 +128,7 @@
 
     &.iui-expand {
-      height: $iui-line-height;
+      height: var(--iui-size-l);
```
```suggestion
      height: var(--iui-size-l);
      min-height: unset;
```

We need to remove the min-height (or change it somehow else) because it looks much larger than it was previously.
![Screen Shot 2022-08-12 at 11 59 42 AM](https://user-images.githubusercontent.com/849817/184395856-8b87ba1c-e659-4bab-87c8-afe9f413e563.png)

FlyersPh9 commented:
The slider icons & drag handle are misaligned and the start / end icons are not receiving the correct color:
![Screen Shot 2022-08-12 at 12 02 26 PM](https://user-images.githubusercontent.com/849817/184396422-7a46634e-3fb7-4880-9f1e-f0ee60660fce.png)

@rohan-kadkol could probably offer a solution as he worked on sliders fairly recently.

FlyersPh9 commented:
Might need to introduce some custom styling for the active state in paginator as we have some really bad contrasting colors.

![Screen Shot 2022-08-12 at 12 06 12 PM](https://user-images.githubusercontent.com/849817/184396858-8a9a0974-ae47-46de-8fc4-e8c08d0ac6e5.png)

FlyersPh9 commented on packages/itwinui-css/src/tag/tag.scss:
```diff
@@ -15,29 +14,28 @@
-  margin-bottom: $iui-component-offset;
-  border-radius: $iui-border-radius-round;
-  padding: 0 $iui-xxs;
+  height: calc(var(--iui-size-s) * 3);
```
```suggestion
  --_iui-tag-height: var(--iui-component-height);
  
  height: var(--_iui-tag-height);
```

Our tags should probably use the new component height variable.

mayank99 replied:
I don't think tags need to always have the same height as inputs/buttons (especially since they have different shapes), so this variable wouldn't make sense here.

FlyersPh9 commented on packages/itwinui-css/src/tree/tree.scss:
```diff
@@ -1,11 +1,14 @@
+); // padding of small borderless button + 1px transparent border
+$iui-expander-button-width: calc(
+  var(--iui-component-height-large) + 2px
+); // width of small borderless button (button is 1:1 ratio)
```
The formatting on this could be improved.  Also I think the math is wrong, our indentations are too large now:

**Before:**
![Screen Shot 2022-08-12 at 12 16 03 PM](https://user-images.githubusercontent.com/849817/184399117-74fc47ab-74d5-4759-ae18-7f177c96e5b5.png)

**After:**
![Screen Shot 2022-08-12 at 12 16 38 PM](https://user-images.githubusercontent.com/849817/184399257-55ed0fa0-0393-439f-ae98-9fa076f90cea.png)

mayank99 replied:
Fixed. There was a typo (`large` -> `small`).

FlyersPh9 commented:
I've looked through every component 1 by 1.  Only things remaining to investigate are `::selection` with status colors (I would _really_ like to keep this) and looking into opacity for scrollbars (I would prefer if we include opacity but I could live without it, just verify the end result looks good on the `color.html` demo page).

rohan-kadkol commented:
> The slider icons & drag handle are misaligned and the start / end icons are not receiving the correct color: ![Screen Shot 2022-08-12 at 12 02 26 PM](https://user-images.githubusercontent.com/849817/184396422-7a46634e-3fb7-4880-9f1e-f0ee60660fce.png)
> 
> @rohan-kadkol could probably offer a solution as he worked on sliders fairly recently.

From a first look, I remember I also faced the problem of different min and max icon colors. I fixed it by checking that one of the old `svg` tag selectors was not accidentally deleted or placed in the wrong scss hierarchy/code block. For the misaligned thumb, I too faced similar problems. It just involved playing around with the values. I guess when including the variables, it might have caused some other problem.

Also, I noticed that lines 41 and 42 were removed from the below screenshot of `sliders.scss`. Not sure, but maybe re-adding that might fix the thumb alignment?

![image](https://user-images.githubusercontent.com/45748283/184400550-13347258-8867-4f93-9260-c749c07823b6.png)

I will take a closer look, play with the values, and update you. Thanks.

mayank99 commented:
@FlyersPh9 Thanks for the detailed review, will try to get these soon.

rohan-kadkol commented on packages/itwinui-css/src/slider/slider.scss:
```diff
@@ -20,7 +19,7 @@ $tick-height: $iui-baseline; // 12px
-    margin-top: $iui-xs;
+    margin-top: var(--iui-size-2xs);
   }
 }
```
To solve the slider icons color problem as mentioned [here](https://github.com/iTwin/iTwinUI/pull/718#issuecomment-1213274552), I think you can include the `iui-slider-color` mixin in the `iui-icon-style` mixin or something similar.

![image](https://user-images.githubusercontent.com/45748283/184411634-299bd352-831e-4ac8-b03a-fcc5b0db35d1.png)

<details>
<summary>Expand to see old code</summary>

#### slider.scss
![image](https://user-images.githubusercontent.com/45748283/184411939-a41b930a-2f1f-4c78-9136-67dac5213a92.png)

#### mixins.scss
![image](https://user-images.githubusercontent.com/45748283/184412103-f5372f3d-e016-460c-86e9-965564c50334.png)

#### mixins.scss
![image](https://user-images.githubusercontent.com/45748283/184412201-76d2e65b-31c6-4dac-938a-1978707a60d0.png)

</details>

mayank99 replied:
That is a specifically a change we don't want to make. The `iui-icon-style` mixin already has too many side effects (hence https://github.com/iTwin/iTwinUI-icons/pull/43) considering how widely it is used.

But I think we could `@include iui-icon-color` in addition to `iui-icon-style` right here in this rule (after line 21).

rohan-kadkol commented on packages/itwinui-css/src/slider/slider.scss:
```diff
@@ -38,9 +37,10 @@ $tick-height: $iui-baseline; // 12px
+  }
 }
 
 @mixin iui-slider-thumb {
```
I don't know if this is important, but one small thing I noticed was that this property might be increasing the size of the thumb from `14x14` to `16x16`. I think this is part of [Leah's PR](https://github.com/iTwin/iTwinUI/pull/661) and might be a needed property, but just mentioning what I noticed. Because this property was not there when I working on the [vertical slider PR](https://github.com/iTwin/iTwinUI/pull/707)

rohan-kadkol replied:
Oops, looks like it selected the wrong lines. I was referring to this property
![image](https://user-images.githubusercontent.com/45748283/184420793-112c7b67-2409-4356-85cc-a642fce057ea.png)

mayank99 replied:
What property?

rohan-kadkol replied:
`box-sizing: content-box` on line 48. This was just a small thing I noticed, so I thought I'd just mention it.

mayank99 replied:
I was looking into this and didn't notice any difference between the size of the thumb in the [live version](https://itwin.github.io/iTwinUI/slider.html) and [current version](https://itwin.github.io/iTwinUI/718/slider.html). They are both 18x18 (16 + 1px border on both sides).

rohan-kadkol replied:
Ohh, I meant when I removed the `box-sizing` property, I noticed that the size of the thumb went from `16x16` to `14x14`. When I was working on the [vertical slider PR](https://github.com/iTwin/iTwinUI/pull/707), this property was not there. I think it came from [Leah's PR](https://github.com/iTwin/iTwinUI/pull/661) done after I worked on mine.

[Screencast from 08-17-2022 01:12:32 AM.webm](https://user-images.githubusercontent.com/45748283/185040275-20d7fdd6-a4a0-41bf-b20b-5f05e8141ff3.webm)

mayank99 replied:
Well yeah, it's not meant to be 14x14 so `content-box` was added to restore the size back to 16x16.

rohan-kadkol replied:
**[RESOLVED]** Ohh, great. Was just confirming

rohan-kadkol commented on packages/itwinui-css/src/slider/slider.scss:
```diff
@@ -102,30 +109,34 @@ $tick-height: $iui-baseline; // 12px
+    top: var(--iui-size-s);
   }
 
   .iui-slider-thumb {
```
For the thumb positioning problem mentioned [here](https://github.com/iTwin/iTwinUI/pull/718#issuecomment-1213274552), I think it is because `$track-height + 1` is producing an unexpected result in the actual webpage's CSS.

#### Selected lines (slider.scss):

![image](https://user-images.githubusercontent.com/45748283/184423415-63ccd849-e33e-4a54-882c-f76491fdc662.png)

#### Declaration of $track-height (slider.scss):

![image](https://user-images.githubusercontent.com/45748283/184423711-7854f377-0d40-4991-98a1-3b82fbc49fe5.png)

#### What appears in the browser:

![image](https://user-images.githubusercontent.com/45748283/184423872-1ee4f275-9787-4114-bfa0-2b81033bc0a0.png)

Here there is an expected 1 that is causing a problem. It works fine when we replace it in the browser with the calculated value of 5px.

mayank99 replied:
Good catch, should be fixed now

veekeys commented on packages/itwinui-css/src/breadcrumbs/breadcrumbs.scss:
```diff
@@ -33,20 +34,20 @@
     @include iui-anchor;
     box-sizing: border-box;
-    line-height: $iui-component-height;
+    line-height: var(--_iui-breadcrumbs-height);
```
sure we wanna have `line-height` with value of `100%` ?

mayank99 replied:
but this is the same as before

veekeys replied:
Well no, previously it was using `iui-component-height` variable values for `line-height`, now you are using param of the mixin value, which is passed as `100%` when breadcrumbs mixin is used in header.

mayank99 replied:
Reverted 👍

veekeys commented on packages/itwinui-css/src/anchor/anchor.scss:
```diff
@@ -50,13 +49,12 @@
   @include iui-focus($color: var(--iui-color-foreground-#{$status}), $offset: 1px);
   color: var(--iui-color-foreground-#{$status});
-  -webkit-tap-highlight-color: rgba(var(--iui-color-foreground-#{$status}-rgb), var(--iui-opacity-5));
   @if $status != primary {
```
why highlight color was removed?

mayank99 replied:
because we don't/didn't have the transparency variables (and because jon was ok with removing it). i think i might add it back though

mayank99 replied:
added back

veekeys commented on packages/itwinui-css/backstop/tests/information-panel.html:
```diff
@@ -1310,7 +1318,8 @@ <h1>Information panel</h1>
                 value="2021"
-                class="iui-input iui-small"
+                class="iui-input"
+                data-iui-size="small"
```
all of that could have been done separately...

mayank99 replied:
I think these changes came from Jon's commits. This was a side effect of removing the component-height variables and aligning the logic with buttons.

veekeys replied:
Cause then we have few more components, which also use class modifiers versus data attribute modifiers.. Sounds weird this change was required because of some variables change :/

veekeys reviewed (commented):
I havent run any of that locally and verified values, would like to firstly see the new visual images, cause it seems everything will be different.

Something I noticed and I really dont like with the "new" variables is that the naming now is so broad, that the readability is so much harder (I am not talking about values here). Some of the examples

- `icon-color` -> `foreground-n`
- `text-color` -> `foreground-n`
- `background-overlay` ->`background-hover` (this variable is also used in multiple places and not even for hover, so hard to understand if its naming issue or something else)
- `font-size-title` -> `font-size-4` (this one is pretty cruel, previously it was much better to guess what to use where we were reusing some of the common naming for different sizes)
- even the baseline change, I do feel having a baseline variable would be so much better

Some of these could be fixed by just adding new variables with duplicated values and some of them could be revisited, cause this naming with random numbers as suffix does not work well in every case as in elevation (which is the only one we have discussed and decided to go with simple numbering system).

These are just some of the thoughts, but the biggest concern is how our apps gonna migrate to the new iTwinUI version, cause I have no idea how they could verify that no layouts are broken or colors arent too different (knowing some of the apps are in the middle of migration to iTwinUI). Of course, fact is, barely any app has visual tests to ensure changes in their UI is acceptable.

Another really important thing is to have all other changes prior to this published as separate dev version, so we could see, how this single commit is influencing things in react side, also we might want to take some of the existing products and try to migrate them to our `alpha` release.

Dont forget to document all of the changes, especially variables replacement.

In general, good job. Im trying to be positive, cause all of that is leading us to something really nice and easy to use, right :)

bentleyvk commented on packages/itwinui-css/src/tabs/pill.scss:
```diff
@@ -1,34 +1,43 @@
+      background-color: var(--iui-color-background-transparent-hover);
+    }
+
+    &.iui-active:hover {
```
It doesn't seem to work on regular tab. But I am more of a fan of having colorful hover states as before.

mayank99 replied:
what doesn't work? this code is for pill tabs

bentleyvk replied:
![tab-v1-hover](https://user-images.githubusercontent.com/36186912/187159242-0da05b8e-7dd7-4ada-899e-0e7f72eba72f.gif)

mayank99 replied:
Fixed

bentleyvk commented on packages/itwinui-css/src/textarea/textarea.scss:
```diff
@@ -1,12 +1,11 @@
   @include iui-input;
   resize: vertical;
-  min-height: $iui-component-height;
+  overflow-y: overlay;
```
?

mayank99 replied:
this file doesn't exist anymore

bentleyvk reviewed (commented):
I checked the code only, need to check visuals later.
The main concerns of mine:
 - foreground changes to background
 - previously we had variables that was clear for what it is like text-muted, font-size-heading, and now there is only number postfix that doesn't tell anything to the users
 - This one looks odd `--iui-color-background-transparent-hover`

mayank99 commented on packages/itwinui-variables/src/typography.scss:
```diff
@@ -15,6 +15,6 @@
   --iui-font-weight-light: 300;
   --iui-font-weight-normal: 400;
-  --iui-font-weight-semibold: 500;
+  --iui-font-weight-semibold: 600;
```
it's the correct value
https://github.com/iTwin/iTwinUI/blob/3b9b327ed8cf84db87dc1d61d356445cd5b051ba/packages/itwinui-css/src/style/typography.scss#L30

mayank99 commented on packages/itwinui-variables/src/sizes.scss:
```diff
@@ -18,3 +18,10 @@
   --iui-border-radius-1: 4px;
```
Not sure I understand what's scary or confusing. I changed 3px -> 4px

mayank99 commented on packages/itwinui-css/src/utils/icon/icon.scss:
```diff
@@ -1,56 +1,51 @@
+
+$iui-icon-sizes: (
+  s: var(--iui-size-s),
+  // 12px
```
Looks like our pre-commit hook puts any comments that are at the of the line into the next line.

I've moved it into the previous line.

mayank99 commented on packages/itwinui-css/src/tooltip/tooltip.scss:
```diff
@@ -1,26 +1,25 @@
   overflow-wrap: break-word;
   box-sizing: content-box;
-  padding: round($iui-baseline * 0.33) $iui-s;
+  padding: calc(var(--iui-size-s) * 0.33) var(--iui-size-xs);
```
This was a find-and-replace. I kept the same math as before

mayank99 commented on packages/itwinui-css/src/tooltip/tooltip.scss:
```diff
@@ -1,26 +1,25 @@
   background-color: rgba(0, 0, 0, var(--iui-opacity-3));
   color: var(--iui-color-foreground-accessory);
-  border: 1px solid rgba(var(--iui-color-foreground-accessory-rgb), var(--iui-opacity-4));
+  border: 1px solid rgba(255, 255, 255, var(--iui-opacity-4));
```
it was not changing across themes previously too.

`-foreground-accessory` is just a fancy name for "white"
https://github.com/iTwin/iTwinUI/blob/3b9b327ed8cf84db87dc1d61d356445cd5b051ba/packages/itwinui-variables/src/colors.scss#L25

mayank99 commented on packages/itwinui-css/src/button/button.scss:
```diff
@@ -1,33 +1,31 @@
   gap: var(--_iui-button-gap);
-  height: var(--_iui-button-height);
+  min-height: var(--_iui-button-min-height);
+  min-width: var(--_iui-button-min-height);
```
this makes the icon-button square (same as #712)

veekeys replied:
Shouldnt it make it square with just having set icon size + padding...?
And I am not sure if `min-height` is better than `height` ? doesnt it allow the button to be larger than we expect it to be then?

mayank99 replied:
setting just one expected value sounds a lot simpler than relying on icon size + padding

also, `min-height` makes sense because we want the height to increase if the content inside is larger for whatever reason. Setting fixed `height` should generally be avoided

mayank99 commented on packages/itwinui-css/src/avatar/avatar.scss:
```diff
@@ -37,11 +36,11 @@
     text-decoration: none;
     text-transform: uppercase;
-    background-color: $iui-color-dataviz-celery;
+    background-color: hsl(72, 51%, 56%);
```
yeah, because those colors should not really be used anywhere. Even for `avatar` we should find a better default color

veekeys replied:
This might be big.. We have whole section about use cases for those colors.. Was this verified with the UX team...?

mayank99 commented on packages/itwinui-css/src/badge/badge.scss:
```diff
@@ -1,27 +1,23 @@
-$iui-badge-horizontal-padding: $iui-xs * 1.5;
-
-$iui-badge-text-color: rgba($iui-color-foreground-body--light, $iui-opacity-3--light);
+$iui-badge-background-positive: hsl(96, 45%, 78%); // foreground-positive--dark
```
we don't have `--dark` or `--light` versions of variables anymore so i hardcoded it, because currently our badges are supposed to look the same across themes.

in the future, i think we want the `soft-background` colors to be themeable so we might be able to use those

mayank99 commented on packages/itwinui-css/src/blockquote/blockquote.scss:
```diff
@@ -1,23 +1,22 @@
   text-align: justify;
-  box-shadow: $iui-elevation-2;
-  border-radius: $iui-border-radius;
+  box-shadow: var(--iui-shadow-1);
```
this was done in #675 because it just makes more sense 🤷

mayank99 commented on packages/itwinui-css/src/button/borderless.scss:
```diff
@@ -21,16 +19,16 @@
 
 @mixin iui-button-borderless-active {
-  --_iui-button-background-color: rgba(var(--iui-color-foreground-primary-rgb), var(--iui-opacity-6));
+  --_iui-button-background-color: var(--iui-color-background-subtle-primary);
```
well, it makes sense as it's used for the background

mayank99 commented on packages/itwinui-css/src/button/default.scss:
```diff
@@ -37,11 +36,7 @@
-    rgba(var(--iui-color-foreground-primary-rgb), var(--iui-opacity-6))
-  ),
-    linear-gradient(var(--iui-color-background-1), var(--iui-color-background-1));
+  --_iui-button-background-color: var(--iui-color-background-subtle-primary);
```
the `-background-subtle` variables are meant to replace the gradient hacks

mayank99 commented on packages/itwinui-css/src/color-picker/color-picker.scss:
```diff
@@ -73,14 +72,10 @@ $iui-hover-box-shadow: 0 0 1px $iui-xxs + 1 rgba(var(--iui-color-foreground-body
-  min-width: $iui-l;
-  gap: $iui-s;
-
-  button[data-iui-variant='borderless'] {
```
not needed anymore because our borderless button is now square.

mayank99 commented on packages/itwinui-css/src/date-picker/date-picker.scss:
```diff
@@ -135,7 +128,7 @@ $iui-date-picker-today-circle-size: 32px;
 
   &::before {
-    border-color: rgba(var(--iui-color-foreground-accessory-rgb), var(--iui-opacity-4));
+    border-color: rgba(255, 255, 255, var(--iui-opacity-4));
```
see https://github.com/iTwin/iTwinUI/pull/718#discussion_r948123325

mayank99 commented on packages/itwinui-css/src/textarea/textarea.scss:
```diff
@@ -1,12 +1,11 @@
   @include iui-input;
   resize: vertical;
-  min-height: $iui-component-height;
+  overflow-y: overlay;
```
this file doesn't exist anymore

mayank99 commented on packages/itwinui-css/src/slider/slider.scss:
```diff
@@ -38,9 +38,10 @@ $tick-height: $iui-baseline; // 12px
 @mixin iui-slider-rail {
   position: absolute;
-  width: 100%;
-  height: $track-height;
```
This code is repeated down below in the `-horizontal` mixin. I think there was a bad merge in the v1 branch when vertical slider was added to the main branch.

mayank99 commented on packages/itwinui-css/src/menu/menu.scss:
```diff
@@ -162,22 +162,26 @@ $iui-active-outline: thin solid var(--iui-color-foreground-primary);
 
 @mixin iui-menu-content {
-  padding: $iui-baseline ($iui-component-padding-horizontal + 1px);
+  padding-block: var(--iui-size-xs);
```
For readability I guess. Jon made these changes.

mayank99 commented on packages/itwinui-css/src/input/input.scss:
```diff
@@ -1,70 +1,76 @@
-    border-color: var(--iui-color-background-disabled);
+    --_iui-input-background-color: var(--iui-color-background-disabled);
+    --_iui-input-border-color: var(--iui-color-background-disabled);
+    --_hover-color: var(--iui-color-background-disabled); // RENAME
```
I don't know why it says rename, this variable has been around for 6 months

veekeys replied:
> > > * `icon-color` -> `foreground-n`
> > > * `text-color` -> `foreground-n`
> 
> > Having aliases for these sounds nice, but it just gets complicated with the different variations of icons and text, and in some cases the variables start getting misused. Whereas with a single foreground scale, you only need to look at the documentation once and it just makes sense after that

Previously I knew, if I need to style icon, i use `icon-` variables, if I need to apply styling to text, i use `text-` variables. I dont think there were anymore variations of these groups. I think we even had a little different opacity for fill of icon and for text color at some point.

Can I see documentation, so it would make sense to me?

> 
> > > * `font-size-title` -> `font-size-4` (this one is pretty cruel, previously it was much better to guess what to use where we were reusing some of the common naming for different sizes)
> 
> > i don't agree with this in particular. previously the font-size variables seemed completely random to me, whereas now, `1` is the base font size, `0` is the smaller one, and `2`, `3`, etc are progressively getting larger.

So how I am supposed to know which of the number to use if I want to have a title of the section? Or which number is the `default` for general text in my app?
Shouldnt we at least start numbering from `1` so this is at least synced in some kind of a way with heading tags of html?
Does this mean our react components will become `FontElement2` too?

> 
> > > * even the baseline change, I do feel having a baseline variable would be so much better
> 
> > how does having a baseline variable help? we never followed the baseline grid anyway, and now it's just part of our normal spacing scale.

I know products have started to adopt baseline vs horizontal spacing recommendations after lots of years of pushing this, so they have their padding set to use `baseline` and `space` variables accordingly. So having an easier replace of the variables might look as an easier transition to the major release.
And in different variable it might be easier to explain, why they should use always `12px` instead of `16px` for vertical default spacing.

bentleyvk replied:
Maybe rename it to meet our standards because it will never be renamed.

mayank99 replied:
Renamed

mayank99 commented on packages/itwinui-css/src/date-picker/date-picker.scss:
```diff
@@ -78,47 +77,41 @@ $iui-date-picker-today-circle-size: 32px;
 @mixin iui-calendar-day-base-hover {
   &:hover {
-    font-weight: $iui-font-weight-semibold;
-    color: var(--iui-color-foreground-primary);
```
There was some large changes made in date-picker to make the hover states simpler.

mayank99 commented on packages/itwinui-css/src/date-picker/date-picker.scss:
```diff
@@ -78,47 +77,41 @@ $iui-date-picker-today-circle-size: 32px;
-    color: var(--iui-color-foreground-primary);
-  }
-
-  &:hover::before {
```
There was some large changes made in date-picker to make the hover states simpler.

mayank99 commented on packages/itwinui-css/src/date-picker/date-picker.scss:
```diff
@@ -205,29 +198,25 @@ $iui-date-picker-today-circle-size: 32px;
-      background-color: rgba(var(--iui-color-foreground-primary-rgb), var(--iui-opacity-5));
-      color: var(--iui-text-color);
-
-      &::before {
```
There was some large changes made in date-picker to make the hover states simpler.

mayank99 commented on packages/itwinui-css/src/file-upload/file-upload.scss:
```diff
@@ -54,27 +53,23 @@
   }
 
   &.iui-drag > .iui-content {
-    background: linear-gradient(
```
the `-background-subtle` variables are meant to replace the gradient hacks

mayank99 commented on packages/itwinui-css/src/location-marker/me.scss:
```diff
@@ -9,18 +8,23 @@
-    rgba(var(--iui-color-foreground-primary-rgb), var(--iui-opacity-5))
-  );
+
+  &::before {
```
This pseudo-element allows us to specify opacity

mayank99 commented on packages/itwinui-css/src/menu/menu.scss:
```diff
@@ -162,22 +162,26 @@ $iui-active-outline: thin solid var(--iui-color-foreground-primary);
+  padding-block: var(--iui-size-xs);
+  padding-inline: calc(var(--iui-size-m) + 1px);
+  min-height: var(--iui-component-height);
+  display: grid;
```
for centering. the height changed a little bit so it was looking off center

mayank99 commented on packages/itwinui-variables/src/themes/light.scss:
```diff
@@ -57,4 +58,9 @@ $opacities: (
   --iui-color-background-border: #{hex(map.get($base-colors, background-5))};
+
+  // Only HSL channels, for customizing transparency
+  --iui-color-background-1-hsl: #{get-hsl(map.get($base-colors, background-1))};
```
the plan is to use these in more places

veekeys replied:
Now or after v1 release? if the latter, we dont need them for this release

mayank99 replied:
using these in a few more places now

mayank99 commented on packages/itwinui-css/src/tile/tile.scss:
```diff
@@ -240,18 +234,20 @@
   &[data-iui-active='true'] {
-    --_iui-button-text-color: var(--iui-color-foreground-primary);
-    --_iui-button-background-color: rgba(var(--iui-color-foreground-primary-rgb), var(--iui-opacity-6));
+    // background-color: var(--iui-color-foreground-primary);
```
removed

mayank99 commented:
> I havent run any of that locally and verified values, would like to firstly see the new visual images, cause it seems everything will be different.

You can check the deploy preview: https://itwin.github.io/iTwinUI/718

> * `icon-color` -> `foreground-n`
> * `text-color` -> `foreground-n`

Having aliases for these sounds nice, but it just gets complicated with the different variations of icons and text, and in some cases the variables start getting misused. Whereas with a single foreground scale, you only need to look at the documentation once and it just makes sense after that

> * `background-overlay` ->`background-hover` (this variable is also used in multiple places and not even for hover, so hard to understand if its naming issue or something else)

yeah i don't think the name makes sense here

> * `font-size-title` -> `font-size-4` (this one is pretty cruel, previously it was much better to guess what to use where we were reusing some of the common naming for different sizes)

i don't agree with this in particular. previously the font-size variables seemed completely random to me, whereas now, `1` is the base font size, `0` is the smaller one, and `2`, `3`, etc are progressively getting larger.

> * even the baseline change, I do feel having a baseline variable would be so much better

how does having a baseline variable help? we never followed the baseline grid anyway, and now it's just part of our normal spacing scale.

elephantcatdog commented on packages/itwinui-css/backstop/tests/index.html:
```diff
@@ -68,7 +68,10 @@
     <style>
       @import url("./assets/demo.css");
+      @import url("@itwin/itwinui-variables");
+      @import url("@itwin/itwinui-variables/os.css");
```
Why do you need to import `os.css` specifically if you've already imported `@itwin/itwinui-variables`? Is it because this is a url import?

mayank99 replied:
`@itwin/itwinui-variables` refers to `index.css`

we specifically want to also import `os.css` in `index.html` because there is no javascript theme switcher in here.

elephantcatdog commented on packages/itwinui-css/src/table/table-densities.scss:
```diff
@@ -1,11 +1,10 @@
 // Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 // See LICENSE.md in the project root for license terms and full copyright notice.
-@import '../style/baseline';
 @import '../style/space';
```
`style/space` doesn't exist anymore either.

veekeys commented:
> > > * `icon-color` -> `foreground-n`
> > > * `text-color` -> `foreground-n`
> 
> > Having aliases for these sounds nice, but it just gets complicated with the different variations of icons and text, and in some cases the variables start getting misused. Whereas with a single foreground scale, you only need to look at the documentation once and it just makes sense after that

Previously I knew, if I need to style icon, i use `icon-` variables, if I need to apply styling to text, i use `text-` variables. I dont think there were anymore variations of these groups. I think we even had a little different opacity for fill of icon and for text color at some point.

Can I see documentation, so it would make sense to me?

> 
> > > * `font-size-title` -> `font-size-4` (this one is pretty cruel, previously it was much better to guess what to use where we were reusing some of the common naming for different sizes)
> 
> > i don't agree with this in particular. previously the font-size variables seemed completely random to me, whereas now, `1` is the base font size, `0` is the smaller one, and `2`, `3`, etc are progressively getting larger.

So how I am supposed to know which of the number to use if I want to have a title of the section? Or which number is the `default` for general text in my app?
Shouldnt we at least start numbering from `1` so this is at least synced in some kind of a way with heading tags of html?
Does this mean our react components will become `FontElement2` too?

> 
> > > * even the baseline change, I do feel having a baseline variable would be so much better
> 
> > how does having a baseline variable help? we never followed the baseline grid anyway, and now it's just part of our normal spacing scale.

I know products have started to adopt baseline vs horizontal spacing recommendations after lots of years of pushing this, so they have their padding set to use `baseline` and `space` variables accordingly. So having an easier replace of the variables might look as an easier transition to the major release.
And in different variable it might be easier to explain, why they should use always `12px` instead of `16px` for vertical default spacing.

bentleyvk commented on packages/itwinui-css/src/color-picker/color-picker.scss:
```diff
@@ -1,39 +1,38 @@
 @import '../surface/index';
 @import '../button/borderless';
 @import '../utils/index';
 
```
Why color picker border went from darker same color to just gray?

bentleyvk replied:
this one also changed to solid color:
![image](https://user-images.githubusercontent.com/36186912/185600264-f356b0cb-fbcb-4fe2-8936-12fdd0e91a6c.png)

mayank99 replied:
I think these changes are intentional and verified

bentleyvk replied:
For real??? Previously swatches looked better in my opinion.

bentleyvk commented on packages/itwinui-css/src/date-picker/date-picker.scss:
```diff
@@ -78,47 +77,41 @@ $iui-date-picker-today-circle-size: 32px;
-    color: var(--iui-color-foreground-primary);
-    background-color: rgba(var(--iui-color-foreground-primary-rgb), var(--iui-opacity-6));
+    font-weight: var(--iui-font-weight-semibold);
+    background-color: var(--iui-color-background-transparent-hover);
```
I don't like this change from blue to the gray as our UI are black, gray and white, we are lacking more colors and this one is removing what is left.

bentleyvk commented on packages/itwinui-css/src/button/borderless.scss:
```diff
@@ -21,16 +19,16 @@
 
 @mixin iui-button-borderless-active {
-  --_iui-button-background-color: rgba(var(--iui-color-foreground-primary-rgb), var(--iui-opacity-6));
+  --_iui-button-background-color: var(--iui-color-background-subtle-primary);
```
This color is barely visible on darker background e.g. inside info-panel 
![image](https://user-images.githubusercontent.com/36186912/185614385-0fb7012e-adb9-4854-8eb2-d5619d9687f9.png)

mayank99 replied:
Should be fixed now

bentleyvk commented on packages/itwinui-css/backstop/tests/progress-indicator.html:
```diff
@@ -25,9 +25,10 @@
       @import url("@itwin/itwinui-css/css/all.css") layer(itwinui);
       section[id^='demo'] {
-        width: 33vw;
+        width: 35vw;
```
Why changed?

mayank99 replied:
this was done to prevent line wrapping as the sizes are now slightly bigger

bentleyvk commented on packages/itwinui-css/src/skip-to-content/skip-to-content.scss:
```diff
@@ -1,27 +1,26 @@
 // See LICENSE.md in the project root for license terms and full copyright notice.
-@import '../style/index';
 @import '../utils/index';
 
```
Broken
![image](https://user-images.githubusercontent.com/36186912/185616286-53deada4-2995-4738-88df-4a3a68e0ee53.png)

mayank99 replied:
Fixed

bentleyvk commented on packages/itwinui-css/src/slider/slider.scss:
```diff
@@ -20,7 +19,8 @@ $tick-height: $iui-baseline; // 12px
   svg {
     @include iui-icon-style('m');
-    margin-top: $iui-xs;
+    @include iui-icon-color;
```
Is this color change ok?
![image](https://user-images.githubusercontent.com/36186912/185616575-69de0bfe-dd58-4c94-8cc4-6188d952ee6f.png)

rohan-kadkol replied:
I'm not sure, but I think the grey one is the correct color. That is how it looks on the website. It was also grey when I was working on the [vertical slider PR](https://github.com/iTwin/iTwinUI/pull/707)

bentleyvk commented on packages/itwinui-css/src/table/table.scss:
```diff
@@ -154,11 +154,7 @@
-          rgba(var(--iui-color-foreground-primary-rgb), var(--iui-opacity-6))
-        ),
-          linear-gradient(var(--iui-color-background-1), var(--iui-color-background-1));
+        background-color: var(--iui-color-background-transparent-hover);
```
I don't like this color change.

mayank99 replied:
This was done to make our hover states more consistent

bentleyvk commented on packages/itwinui-css/src/tile/tile.scss:
```diff
@@ -1,6 +1,5 @@
 // Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 // See LICENSE.md in the project root for license terms and full copyright notice.
-@import '../style/index';
 @import '../button/index';
```
Idk if this is known but icons background is missing 
![image](https://user-images.githubusercontent.com/36186912/185617135-5a1492c8-ada4-4cb4-874e-7bd0f6d63094.png)

mayank99 replied:
where is that from? i'm not seeing any missing icons

bentleyvk replied:
ok, why background was changed from darker to this, barely visible one? Now it is harder to see that there is an icon in the corner if image is bright and previously we didn't have such problem since icon background was darker.

mayank99 replied:
I'm still not sure what part you're talking about. 

https://itwin.github.io/iTwinUI/718/tile.html and https://itwin.github.io/iTwinUI/tile.html look the same

bentleyvk replied:
I am seeing this in local build. As usual build must be fucked up...

bentleyvk commented on packages/itwinui-css/src/code/codeblock.scss:
```diff
@@ -39,21 +38,21 @@
 
       &:nth-child(even) {
-        background: rgba(var(--iui-color-foreground-body-rgb), 0.02);
+        background: var(--iui-color-background-1-hover);
```
Why on dark theme these got inverted? 
![image](https://user-images.githubusercontent.com/36186912/185617581-d059e9df-c012-48ee-a373-9ed0e3fc9376.png)

mayank99 replied:
Should be fixed now

mayank99 commented on packages/itwinui-css/src/slider/slider.scss:
```diff
@@ -102,37 +110,39 @@ $tick-height: $iui-baseline; // 12px
     cursor: not-allowed;
     background-color: var(--iui-color-background-4);
     border-color: var(--iui-color-background-4);
+    @media (forced-colors: active) {
```
This might have been a bad merge in the v1 branch from when the vertical slider PR was merged into main
https://github.com/iTwin/iTwinUI/blob/3b9b327ed8cf84db87dc1d61d356445cd5b051ba/packages/itwinui-css/src/slider/slider.scss#L121

mayank99 commented on packages/itwinui-css/src/input/input.scss:
```diff
@@ -1,70 +1,76 @@
+  background-color: var(--_iui-input-background-color);
+  border: 1px solid var(--_iui-input-border-color);
+  transition: border-color var(--iui-duration-1) ease-out;
+  box-sizing: border-box;
```
I think these were again done by Jon. I've removed the `box-sizing` property because it's now set globally, but I've kept the removal of media query as it's not necessary.

mayank99 commented:
> Might need to introduce some custom styling for the active state in paginator as we have some really bad contrasting colors.
> 
> ![Screen Shot 2022-08-12 at 12 06 12 PM](https://user-images.githubusercontent.com/849817/184396858-8a9a0974-ae47-46de-8fc4-e8c08d0ac6e5.png)

I'm now using `foreground-primary-hsl` + `opacity-6` for button active states, so this is fixed now (behaving same as before).

mayank99 commented on packages/itwinui-css/src/expandable-block/block.scss:
```diff
@@ -14,11 +13,11 @@
     display: flex;
     align-items: center;
-    padding: $iui-baseline $iui-sm;
+    padding: var(--iui-size-s) var(--iui-size-s);
```
done

mayank99 commented on packages/itwinui-css/src/color-picker/color-picker.scss:
```diff
@@ -1,39 +1,38 @@
-$iui-hover-box-shadow: 0 0 1px $iui-xxs + 1 rgba(var(--iui-color-foreground-body-rgb), var(--iui-opacity-5));
+$iui-inset-box-shadow: inset 0 0 0 1px var(--iui-color-foreground-5);
+$iui-active-box-shadow: 0 0 0 var(--iui-size-3xs) var(--iui-color-foreground-primary);
+$iui-hover-box-shadow: 0 0 0 var(--iui-size-3xs) var(--iui-color-foreground-4); // 0 0 1px calc(var(--iui-size-3xs) + 1px) var(--iui-color-background-1-hover);
```
i think this is intentional. When pairing with Jon, we changed the visuals a little bit

mayank99 commented on packages/itwinui-css/src/header/header-buttons.scss:
```diff
@@ -163,16 +163,19 @@
 
-  @at-root [aria-current] &:hover {
-    background-color: rgba(var(--iui-color-foreground-primary-rgb), var(--iui-opacity-6));
+  &:hover::before {
```
this was needed because we didn't have transparency variables. i've now reverted this

rohan-kadkol commented:
> > I'm seeing a visual bug with the dropdown icon in the small select within the information panel demo:
> > ![Screen Shot 2022-08-12 at 10 23 05 AM](https://user-images.githubusercontent.com/849817/184373693-b2e76218-4db1-46d3-860d-98f8e32f5fbc.png)
> 
> @rohan-kadkol Did you get a chance to look at this^?

Sorry for the delay. I am reposting this here just in case someone else knows why this could be happening and how to fix it.

I have been trying to figure out the dropdown menu button for quite some time now. Unfortunately, I haven't found the solution so far. All I notice is that:
- When passing a width to the `span` with class of `iui-input-icon iui-actionable`, it for some reason also included the padding. E.g. when I pass a width of `19px`, the width becomes `1px` because of a `9px` padding on both the left and the right (`9px` + `9px` = `18px` and `19px` - `18px` = `1px` which becomes the width). But when I pass the same width of `19px` in the hosted demo website's same `span`, the box gets the width of `19px`.
![image](https://user-images.githubusercontent.com/45748283/186521039-f8b13468-6100-4e74-9714-7776d43b42a2.png)
![Screenshot from 2022-08-24 16-34-19](https://user-images.githubusercontent.com/45748283/186521817-482cb433-afef-496d-80f7-f929842cab03.png)

- When removing the width attribute from line 27 of icon.scss, for some reason, the span with class automatically gets a width of 16px and looks almost correct.
![image](https://user-images.githubusercontent.com/45748283/186520963-9fd2ceaf-a1ff-4f39-af3d-11ad9307cc8d.png)
![Screenshot from 2022-08-24 16-35-45](https://user-images.githubusercontent.com/45748283/186522010-a39d4ac3-ff85-426f-a186-5c6cc7fa3566.png)

mayank99 commented:
I've fixed all remaining issues and gone through each component one to verify, before updating test images.

Next thing I'm going to document all the variables and create a migration guide to help users switch over from old sass variables. But that shouldn't really hold this PR.

veekeys commented on packages/itwinui-css/src/utils/input-container/input-container.scss:
```diff
@@ -186,10 +186,11 @@
   &.iui-actionable {
     align-items: center;
-    height: 90%;
+    height: calc(100% - 4px);
```
comment for 4px might be useful

mayank99 replied:
I'm subtracting 2px on both sides to avoid overlapping with border/outline

Added comment

veekeys reviewed (commented):
SOme points I noticed while going through images.
- Why `Alert` close icon moved to the right
- Now seems our buttons are smaller height
- disabled checkbox checkmark color changes is connected with increase in wcag score, I assume
- `CodeBlock` title now less height
- `ColorPicker` thumb to choose color (or w/e is called) lost border - IMPORTANT
- Color swatches now have some weird border and colors look blurry compared with previous version (might be images issues though)
- Now seems focus and selected outlines are the same? thats weird
- In iTwinUI_dark_button_Grouped_Borderless_0_demo-grouped-borderless_0_desktop disabled llooks also changed
- Menu hover color looks bad :( (we talked about it)
- Why some screenshots increased width??? i.e. prgoress indicator
- Vertical slider icons look misaligned - IMPORTANT
- Table has lost selected border in between different color
- Why only some od the disabled states colors have changed i.e. radio tile, but not toggle switch (in dark mode)

In general, looks weird, because in most of the places height has become less (seems baseline went back from 12 to 11 :/

mayank99 commented:
> * Why `Alert` close icon moved to the right
> * Now seems our buttons are smaller height

button size changes are intentional

> * disabled checkbox checkmark color changes is connected with increase in wcag score, I assume

I don't think I see a difference. If there is a small difference, it's likely because of transparency changes, which is expected in most places.

> * `CodeBlock` title now less height

i think this is because of the copy button, which is intentional

> * `ColorPicker` thumb to choose color (or w/e is called) lost border - IMPORTANT

fixed

> * Color swatches now have some weird border and colors look blurry compared with previous version (might be images issues though)

i think those color swatch changes are intentional

> * Now seems focus and selected outlines are the same? thats weird

they still look different to me

> * In iTwinUI_dark_button_Grouped_Borderless_0_demo-grouped-borderless_0_desktop disabled llooks also changed

aligning disabled styling for default vs borderless buttons is intentional

> * Why some screenshots increased width??? i.e. prgoress indicator

i increased the width because i noticed some of the sections were wrapping

> * Vertical slider icons look misaligned - IMPORTANT

I don't think this is related to my PR. It looks like the vertical slider spacing could use a bit of cleanup so I've opened a new issue: https://github.com/iTwin/iTwinUI/issues/753

> * Table has lost selected border in between different color

Should be fixed now

> * Why only some od the disabled states colors have changed i.e. radio tile, but not toggle switch (in dark mode)

Not sure I see any changes in radio tile

gretanausedaite reviewed (commented):
Is this color picker border change intentional? 
![image](https://user-images.githubusercontent.com/83585998/187939780-3b42c40b-5420-4f8d-ae02-2a5257cb455b.png)

mayank99 commented:
> Is this color picker border change intentional?

I believe it is, because it was harder to see previously. But I can also revert it if we don't want random changes like this.

FlyersPh9 commented:
> > Is this color picker border change intentional?
>
> I believe it is, because it was harder to see previously. But I can also revert it if we don't want random changes like this.

Here's what the [production color picker](https://itwin.github.io/iTwinUI-react/?path=%2Fstory%2Fcore-colorpicker--advanced) looks like when in light theme & placed in the top left corner in light theme. Pretty much invisible unless you hover it. 😕
![Screen Shot 2022-09-01 at 2 25 52 PM](https://user-images.githubusercontent.com/849817/187986211-a8e691de-29fe-4aac-92b6-33eae87487d0.png) ![Screen Shot 2022-09-01 at 2 27 13 PM](https://user-images.githubusercontent.com/849817/187986412-dd6aea58-17a5-4f4f-9605-039dd7bd22ea.png)

But it seems like we have the same issue when the border is set to black and placed in the bottom left corner.
![Screen Shot 2022-09-01 at 2 29 35 PM](https://user-images.githubusercontent.com/849817/187986820-70db98dc-f69f-4e45-b651-f7da34bd7235.png) ![Screen Shot 2022-09-01 at 2 30 09 PM](https://user-images.githubusercontent.com/849817/187986913-a3136ad5-fd27-4163-b6e5-7bf3f3b99051.png)

gretanausedaite reviewed (approved):
Variables looks good to me.
We should fix ColorPicker border issue in next PR before releasing v1.

bentleyvk commented:
Are you planning to fix these?
https://github.com/iTwin/iTwinUI/pull/718#discussion_r948006908
https://github.com/iTwin/iTwinUI/pull/718#discussion_r947907694

mayank99 commented:
> Are you planning to fix these? [#718 (comment)](https://github.com/iTwin/iTwinUI/pull/718#discussion_r948006908) [#718 (comment)](https://github.com/iTwin/iTwinUI/pull/718#discussion_r947907694)

Somehow missed these because github decided to hide them. Fixed them just now
